src/main.rs

```diff
@@ -58,20 +63,31 @@ impl LanguageServer for Backend {
             .to_file_path()
             .unwrap()
             .with_extension("pdf");
-
-        let output = match typst::compile(world) {
+        let messages: Vec<_> = match typst::compile(world) {
             Ok(document) => {
                 let buffer = typst::export::pdf(&document);
-                fs::write(output_path, buffer).map_err(|_| "failed to write PDF file".to_string())
-            }
-            Err(errors) => {
-                let messages: Vec<_> = errors.iter().map(|error| error.message.as_str()).collect();
-                Err(messages.join("\n"))
+                let _ = fs::write(output_path, buffer)
+                    .map_err(|_| "failed to write PDF file".to_string());
+                vec![]
             }
+            Err(errors) => errors.iter().map(|x| error_to_range(x, world)).collect(),
         };
+        drop(world);
 
         self.client
-            .log_message(MessageType::INFO, format!("{:?}", output))
+            .publish_diagnostics(
+                params.text_document.uri,
+                messages
+                    .into_iter()
+                    .map(|(message, range)| Diagnostic {
+                        range,
+                        severity: Some(DiagnosticSeverity::ERROR),
+                        message,
+                        ..Default::default()
+                    })
+                    .collect(),
+                None,
+            )
```

Comment:
Can you rebase so the only changes left are hover related?

Reply:
sure rebase on your branch?